In [1]:
!pip uninstall langchain
!pip install langchain==0.0.339

     ---------------------------------------- 2.0/2.0 MB 3.2 MB/s eta 0:00:00
  Using cached tenacity-8.2.3-py3-none-any.whl (24 kB)
  Using cached jsonpatch-1.33-py2.py3-none-any.whl (12 kB)
  Using cached dataclasses_json-0.6.4-py3-none-any.whl (28 kB)
  Using cached langsmith-0.0.92-py3-none-any.whl (56 kB)
     ---------------------------------------- 407.9/407.9 kB ? eta 0:00:00
  Using cached typing_inspect-0.9.0-py3-none-any.whl (8.8 kB)
  Using cached marshmallow-3.21.1-py3-none-any.whl (49 kB)
  Using cached annotated_types-0.6.0-py3-none-any.whl (12 kB)
     ---------------------------------------- 1.9/1.9 MB 9.4 MB/s eta 0:00:00
  Attempting uninstall: tenacity
    Found existing installation: tenacity 8.0.1
    Uninstalling tenacity-8.0.1:
      Successfully uninstalled tenacity-8.0.1
  Attempting uninstall: jsonpatch
    Found existing installation: jsonpatch 1.32
    Uninstalling jsonpatch-1.32:
      Successfully uninstalled jsonpatch-1.32


In [3]:
!pip install google-generativeai

  Using cached google_generativeai-0.5.0-py3-none-any.whl (142 kB)
  Using cached google_api_core-2.18.0-py3-none-any.whl (138 kB)
  Using cached google_ai_generativelanguage-0.6.1-py3-none-any.whl (663 kB)
  Using cached google_api_python_client-2.125.0-py2.py3-none-any.whl (12.5 MB)
  Using cached proto_plus-1.23.0-py3-none-any.whl (48 kB)
  Using cached googleapis_common_protos-1.63.0-py2.py3-none-any.whl (229 kB)
  Using cached google_auth_httplib2-0.2.0-py2.py3-none-any.whl (9.3 kB)
  Using cached httplib2-0.22.0-py3-none-any.whl (96 kB)
  Using cached uritemplate-4.1.1-py2.py3-none-any.whl (10 kB)
  Using cached grpcio_status-1.62.1-py3-none-any.whl (14 kB)
     ---------------------------------------- 3.8/3.8 MB 3.6 MB/s eta 0:00:00
  Attempting uninstall: grpcio
    Found existing installation: grpcio 1.54.2
    Uninstalling grpcio-1.54.2:
      Successfully uninstalled grpcio-1.54.2


In [4]:
from langchain.llms import GooglePalm

api_key = 'AIzaSyAXRK813RdaDrvVGggt6NUdj1VQ_d34NDQ'

llm = GooglePalm(google_api_key=api_key, temperature=0.1)

In [5]:

poem = llm("Write a 4 line poem of my love for samosa")
print(poem)

**Oh, samosa, my love**
**You are so delicious, so flaky**
**With your fillings of potatoes and peas**
**I could eat you every day**


In [6]:
essay = llm("write email requesting refund for electronic item")
print(essay)

Dear [Name of Customer Service Representative],

I am writing to request a refund for the [product name] that I purchased on [date]. I am not satisfied with the product because [explain why you are not satisfied].

I have attached a copy of my receipt and a picture of the product. I have also included a link to the product's listing on your website.

I would like to request a full refund for the product. I would also like to request that you send me a replacement product.

I have been a loyal customer of your company for many years, and I have never had a problem with a product before. I am disappointed that I am not satisfied with this product, but I hope that you will be able to resolve this issue for me.

Thank you for your time and consideration.

Sincerely,
[Your Name]


In [7]:

from langchain.chains import RetrievalQA


from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

Now let's load data from Codebasics FAQ csv file

In [9]:
from langchain.document_loaders.csv_loader import CSVLoader

# Specify the file path and the correct encoding
file_path = 'codebasics_faqs.csv'
encoding = 'windows-1252'  # Example: windows-1252 or any other encoding you identified

# Initialize the CSVLoader with the specified encoding
loader = CSVLoader(file_path=file_path, source_column="prompt", encoding=encoding)

# Load the data from the CSV file
data = loader.load()


In [10]:
!pip install python-dotenv==1.0.0
!pip install streamlit==1.22.0
!pip install tiktoken==0.4.0
!pip install faiss-cpu==1.7.4
!pip install protobuf~=3.19.0

  Using cached python_dotenv-1.0.0-py3-none-any.whl (19 kB)


ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'C:\\Users\\Admin\\anaconda3\\Lib\\site-packages\\google\\~upb\\_message.pyd'
Consider using the `--user` option or check the permissions.



  Using cached streamlit-1.22.0-py2.py3-none-any.whl (8.9 MB)
  Using cached Pympler-1.0.1-py3-none-any.whl (164 kB)
     -------------------------------------- 904.0/904.0 kB 2.1 MB/s eta 0:00:00
  Using cached pydeck-0.8.1b0-py2.py3-none-any.whl (4.8 MB)
  Using cached tzlocal-5.2-py3-none-any.whl (17 kB)
  Using cached validators-0.28.0-py3-none-any.whl (39 kB)
  Using cached blinker-1.7.0-py3-none-any.whl (13 kB)
  Using cached rich-13.7.1-py3-none-any.whl (240 kB)
  Using cached altair-4.2.2-py3-none-any.whl (813 kB)
  Using cached GitPython-3.1.43-py3-none-any.whl (207 kB)
  Using cached gitdb-4.0.11-py3-none-any.whl (62 kB)
  Using cached markdown_it_py-3.0.0-py3-none-any.whl (87 kB)
     ---------------------------------------- 1.2/1.2 MB 6.2 MB/s eta 0:00:00
  Using cached tzdata-2024.1-py2.py3-none-any.whl (345 kB)
  Using cached smmap-5.0.1-py3-none-any.whl (24 kB)
  Using cached mdurl-0.1.2-py3-none-any.whl (10.0 kB)
  Attempting uninstall: pygments
    Found existing insta

     ---------------------------------------- 10.8/10.8 MB 5.4 MB/s eta 0:00:00


  Using cached protobuf-3.19.6-cp310-cp310-win_amd64.whl (895 kB)
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.20.3
    Uninstalling protobuf-3.20.3:
      Successfully uninstalled protobuf-3.20.3


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow-intel 2.15.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 3.19.6 which is incompatible.
grpcio-status 1.62.1 requires protobuf>=4.21.6, but you have protobuf 3.19.6 which is incompatible.


Hugging Face Embeddings

In [11]:
pip install -U sentence-transformers==2.2.2

     -------------------------------------- 86.0/86.0 kB 373.0 kB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
     ---------------------------------------- 1.2/1.2 MB 1.7 MB/s eta 0:00:00
     -------------------------------------- 991.5/991.5 kB 2.9 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125960 sha256=f26fadd0c074f676bf3bb3346a9664e83215763ab5359546b8de737a3a8dccf8
  Stored in directory: c:\users\admin\appdata\local\pip\cache\wheels\0a\f5\dd\9d00836c4e9e279c2a59d5b0ab72dafa66cbc626a327c550dd
Successfully built sentence-transformers
Note: you may need to restart the kernel to use updated packages.


In [12]:
!pip install InstructorEmbedding

In [13]:

from langchain.embeddings import HuggingFaceInstructEmbeddings

# Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")

e = instructor_embeddings.embed_query("What is your refund policy?")

.gitattributes:   0%|          | 0.00/1.48k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/270 [00:00<?, ?B/s]

2_Dense/config.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/3.15M [00:00<?, ?B/s]

README.md:   0%|          | 0.00/66.3k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.53k [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.41k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/461 [00:00<?, ?B/s]

load INSTRUCTOR_Transformer
'NoneType' object has no attribute 'cadam32bit_grad_fp32'


C:\Users\Admin\anaconda3\lib\site-packages\bitsandbytes\cextension.py:34: UserWarning: The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers, 8-bit multiplication, and GPU quantization are unavailable.
  warn("The installed version of bitsandbytes was compiled without GPU support. "



max_seq_length  512


In [14]:
len(e)

768

In [15]:

e[:5]

[-0.043898072093725204,
 0.00768554350361228,
 -0.009231901727616787,
 0.02449624426662922,
 0.03359228000044823]

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

Vector store using FAISS

In [16]:
from langchain.vectorstores import FAISS

# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)

# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [17]:

rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

[Document(page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': 'Do you provide any job assistance?', 'row': 11}),
 Document(page_content='prompt: Will this course guarantee me a job?\nresponse: We created a much lighter version of this course on YouTube available for free (click this link) and many people gave us feedback that they were able to fetch jobs (see testimonials). Now this paid course is at least 5x better than the YouTube course which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': 'Will this 

Create RetrievalQA chain along with prompt template 🚀

In [18]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)

We are all set 👍🏼 Let's ask some questions now

In [19]:
chain('Do you provide job assistance and also do you provide job gurantee?')

{'query': 'Do you provide job assistance and also do you provide job gurantee?',
 'result': 'Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters. The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.',
 'source_documents': [Document(page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': 'Do you prov

In [20]:
python V

SyntaxError: invalid syntax (2169472906.py, line 1)

In [21]:
import platform
print(platform.python_version())


3.10.9
